sensor/extrinsic_reflector_based_calibrator/src/extrinsic_reflector_based_calibrator.cpp

```diff
@@ -1280,6 +1324,112 @@ void ExtrinsicReflectorBasedCalibrator::calibrateSensors()
       "The calibrated poses differ considerably with the initial calibration. This may be either a "
       "fault of the algorithm or a bad calibration initialization");
   }
+
+  // for plotter and rviz display
+  output_calibration_distance_error = static_cast<float>(calibrated_2d_distance_error);
+  output_calibration_yaw_error = static_cast<float>(calibrated_2d_yaw_error);
+}
+
+void ExtrinsicReflectorBasedCalibrator::crossValEvaluation(
+  pcl::PointCloud<PointType>::Ptr lidar_points_pcs,
+  pcl::PointCloud<PointType>::Ptr radar_points_rcs)
+{
+  // Evaluate the cross validation metric using SVD
+  double cv_distance_error = 0.0;
+  double cv_yaw_error = 0.0;
+  auto get_yaw_error = [](const Eigen::Vector3d & v1, const Eigen::Vector3d & v2) -> double {
+    return std::abs(std::acos(v1.dot(v2) / (v1.norm() * v2.norm())));
+  };
+
+  if (converged_tracks_.size() > 1) {
+    std::vector<int> indices_vec(converged_tracks_.size());
+    std::iota(indices_vec.begin(), indices_vec.end(), 0);
+    for (std::size_t rotate = 0; rotate < indices_vec.size(); ++rotate) {
```

Comment:
I did not finish reading the cross validation, but I am fairly sure that  this is not the requested algorithm.
As new data comes the "real" cross validation for earlier samples also changes.

If I were to write a little bit more format it would be

cross_validation_error(number_of_samples, tracks)

where number_of_samples < |tracks|

When |tracks| change, also does cross_validation_error(number_of_samples, tracks) for "already computed" number_of_samples. 

Also, you mention random in the code, but there is no random component.

I mentioned before that this is a combinatorial problem, and as such prone to be slow, for which reason we needed an stochastic estimator with a limited budget in the sample size. 
You mentioned that due to the problem size that was not needed but I do not see that in the code as of now.

For example,
If we have 20 samples, and want to compute the cross_validation_error using only 3 samples for calibration,
The number of combinations we need to try is 1140

https://www.calculatorsoup.com/calculators/discretemathematics/combinations.php

Reply:
@knzo25 Thanks for your review. 
As you mentioned, we discussed the `random number` and the `number of samples `before. 
However, there are several reasons why I didn't use the random number and the number of samples to implement the cross-validation. 

Currently, the implementation is a leave-one-out cross-validation method, which is also a kind of cross-validation. 
I implemented this method for several reasons. First of all, it won't increase the number of combinations too fast because the number of combinations grows linearly, which always equals the (converged_track - 1). Also, it always takes the new converged track into the calculation. 

For example, there are two cases that I think probably can describe what you mentioned above. 
`Example 1`
Let's say we set the number of samples as 20. If `number of samples >  converged tracks`, we do the leave-one-outcross-validationn method. Now, if `number of samples < converged tracks`, for instance, `converged tracks = 21
`,  we randomly select 20 tracks for calibration and validate 1 track. This will give us 21 different combinations. If the converged tracks = 22 now, we will have 231 different combinations. And the number of combinations will grow super fast when the number of converged tracks grows. And of course, this method will do the `cross_validation_error(number_of_samples, tracks) for "already computed" number_of_samples.` that you mentioned above with an additional large number of combinations. 
With this example, the computation time will grow too fast when the converged track grows. 

If `Example 1` isn't the one you mentioned, then probably Example 2 which elaborates below will be the one you mentioned.
 
`Example 2`
Let's say we fix the number of samples as 20. When the number of tracks is equal to 21, 22, ... 100, we always only calibrate for "already computed" number_of_samples, which is the first 20 samples, and compare with the rest. In this kind of method, the cross-validation will always change, but the calibration result is always the same as we always used the same samples to calculate the transformation matrix. 
With this example, I think it can tell the user that it is time to stop if it is converged. However, what should we do for the case when the number of tracks is from 5 to 19?

Maybe the two examples above are not the case what you mean, then we can discuss what kind of cross-validation you want tomorrow :)

Reply:
@vividf 
The problem with your approach is that the shape of the graph depends on the order the data is taken, as such is does not allow us to see what we need. Earlier points in the graph are highly biased and must be revisited when we need more data

Reply:
Thanks. The new cross-validation algorithm is implemented in commit https://github.com/tier4/CalibrationTools/pull/145/commits/d2299288a6f82602bcc8eead60a504072958ca67. 
![image](https://github.com/tier4/CalibrationTools/assets/146902905/28343bdd-902d-4d55-ba04-d3e42c9926f4)

Reply:
This graph is quite different from the one you sent me earlier. Do the graphs correspond to different bags/sensors?
![image](https://github.com/tier4/CalibrationTools/assets/4712270/e1e35ebc-5ac0-482d-8fee-84ff64e0bdc5)

@vividf 
Since there have been so many changes, I will have to read and run it all again. When you finish, press the corresponding button to trigger another review process

Reply:
@knzo25 
Yes, I also noticed it and I checked it again. The `original bag (download from the drive)` is the same one. However, the earlier one is using the bag only record /calibration metrics when I am doing experiments. I think at that time probably I clicked the add_pair too fast or I played the "original bag" too fast which caused the std to become big. And actually, when the selection number became 9, the result looked similar to the current one (which only have huge difference of std in` number of selection 3`" 

After I use the code that I generated for the `experiment bag `again and clicked the button carefully, the graph is same as the current result.

Reply:
@knzo25 Thanks a lot for taking time to review again!

Reply:
@knzo25 I can kind of reproduce the graph without adding the first pair (when the person raises the hand). Probably, when I recorded the experiment bag before I also didn't add the pair.

![image](https://github.com/tier4/CalibrationTools/assets/146902905/e691e662-60c6-4e58-8491-595cfdf1d6b3)